### ISSUE TITLE
allow $schema in declaration

### ISSUE THREAD

**Issue by simonkowallik** [F5Networks/f5-appsvcs-extension#173]
##### Do you already have an issue opened with F5 support?
No

##### ISSUE TYPE
 - Feature Idea

##### AS3 BUILD/ VERSION
```
3.14.0
```

##### BIGIP VERSION
```
14.1.2
```

##### SUMMARY
`$schema` can be used to validate the declaration in VS Code as per [the User Guide](https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/userguide/validate.html).
When `$schema` is not removed from the delcaration before submitting it, the AS3 iControlLX responds with an error (see below).
```
{
    "id": "778f21e1-c627-4bdd-bad7-49135c6bfcda",
    "results": [
        {
            "code": 422,
            "errors": [
                ": propertyName \"$schema\" should match pattern \"^[A-Za-z][0-9A-Za-z_]*$\""
            ],
            "declarationFullId": "",
            "message": "declaration is invalid"
        }
    ],
    "declaration": {}
}
```

##### STEPS TO REPRODUCE
Leave `$schema` in the delcaration and submit it.

##### EXPECTED RESULTS
AS3 iControlLX extension ignores `$schema` and the declaration succeeds

##### ACTUAL RESULTS
see above

**Comment by markchart** [comment]
This is _not_ a valid RFE.  No property named `$schema` should appear in an AS3 declaration.  The [`$schema` keyword is used only in a JSON Schema file](https://json-schema.org/understanding-json-schema/reference/schema.html) (which tells how to validate a JSON document like a declaration).

(Possibly the User Guide page linked by @simonkowallik could be extended with even more tutorial explanations about this, but the info already on that page is valid and the current behavior of AS3 is correct.)

**Comment by dstokesf5** [comment]
Declarative Onboarding is setup to ignore a "$schema" property in submitted declarations. I though AS3 was as well, but I am not seeing anything in the AS3 code base.

**Comment by markchart** [comment]
I suggest it should _not_ be ignored (I don't know what's going on with DO, but perhaps some developer wasn't fully expert on this stuff).  AS3 should reject declarations with bogus `$schema` properties just like it rejects those with other bogus properties.  Otherwise customers may think they are getting some benefit which they are not actually getting and/or customers may mess up their F5 devices by submitting erroneous declarations.

Notice that the original poster clearly does not understand the difference between schema and declaration and thinks that he is validating his declaration with Visual Studio when that is _not_ what is happening.  If AS3 were not giving him a perfectly valid error message, he would possibly soldier on in complete ignorance until he discombobulated his network with an erroneous declaration.

I see that [Step 7 on the docs page linked by @simonkowallik is bogus](https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/userguide/validate.html).  That should be removed.

**Comment by jsevedge** [comment]
Typically including $schema in what is POSTed is not what a service would expect, but if the service (AS3/DO/TS) wants to silently allow for that (even if not called out in examples, etc.) to enable customers using VS Code to have a better document editing experience when using IntelliSense (following this guidance: https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings) that would appear to be a PM decision to either 1) silently allow it like DO does or 2) explicitly block it and add a troubleshooting note calling out that is not allowed and to instead use one of the other mechanisms called out in the aforementioned link to let VS code know where the schema resides it should valid the document against.

**Comment by markchart** [comment]
Since @jsevedge points out that MS seems to have embraced-and-extended**™** JSON Schema for Visual Studio, then I suggest implementing his option (1) in the following manner:  To the proper AS3 schemas (those for class=AS3 and class=ADC at least) add a couple of lines to validate a property named `$schema` as a string so AJV won't throw an error on it, with a new AJV custom keyword `f5remove` that deletes that property before anything else in AS3 chokes on it:

```
"properties": {
    "$schema": {
        "title": "schema reference",
        "description": "A URI reference to a JSON Schema to be used to validate this document may appear here to help a programmer's editor or automation tool.  Note that AS3 does NOT use this value for any purpose!",
        "type": "string",
        "f5remove": true
    },
```

The custom keyword would be straightforward to define similar to this example, which I have shamelessly derived from [AJV Issue #715](https://github.com/epoberezkin/ajv/issues/715#issuecomment-368939233).  Note that someone would have to define suitable similar code for AS3:

```
const ajv = new Ajv({ allErrors: true, jsonPointers: true });

function isEmptyObject(obj) {
  if (obj == null || typeof obj !== "object") return false;
  for (const prop in obj) {
    return false;
  }
  return true;
}

function clearEmptyPath(obj, path) {
  const parentObjects = [];
  let _obj = obj;
  for (let i = 0; i < path.length - 1; i++) {
    parentObjects.push(_obj);
    _obj = _obj[path[i]];
  }

  for (let i = path.length - 2; i >= 0; i--) {
    if (isEmptyObject(parentObjects[i][path[i]])) {
      delete parentObjects[i][path[i]];
    }
  }
}

ajv.addKeyword("f5remove", {
  type: boolean,
  errors: false,
  modifying: true,
  validate(value, data, parentSchema, currentDataPath, parentDataObject, propertyName, rootObject) {
     delete parentDataObject[propertyName];
      // parent object is empty?
      if (isEmptyObject(parentDataObject)) {
        clearEmptyPath(rootObject, pointer.parse(currentDataPath));
      }
    }
    return true;
  }
});
```

**Comment by dstokesf5** [comment]
I have added AUTOTOOL-2048 to our internal product backlog to ignore the $schema property like Declarative Onboarding already does.

**Comment by dstokesf5** [comment]
This was resolved in the 3.24.0 release.
